### PR TITLE
add go happy-case unit testing

### PR DESCRIPTION
**Pull request by pongad**
This commit adds a change to make this easier:

The "zero value" of a message type *T is &T{} instead of nil.

This necessitates a small change in iterator implementation,
though the behavior should stay the same.

### CODE REVIEW

File: src/test/java/com/google/api/codegen/testdata/go_main_library.baseline [hunk 1]
```diff
@@ -688,13 +688,13 @@ func (it *BookIterator) PageInfo() *iterator.PageInfo {
 
 // Next returns the next result. Its second return value is iterator.Done if there are no more
 // results. Once Next returns Done, all subsequent calls will return Done.
-func (it *BookIterator) Next() (*librarypb.Book, error) {
-    if err := it.nextFunc(); err != nil {
-        return nil, err
+func (it *BookIterator) Next() (item *librarypb.Book, err error) {
```
Comment (jba) [inline src/test/java/com/google/api/codegen/testdata/go_main_library.baseline]: Can you revert to the previous implementation? They seem identical in behavior (unless I'm missing something) and the original is more idiomatic Go.

Comment (jba) [inline src/test/java/com/google/api/codegen/testdata/go_main_library.baseline]: Your change to the generator is fine by me, but you could also write the code this way:

```
func (it *BookIterator) Next() (*librarypb.Book, error) {
    var item *libraryp.Book
    if err := it.nextFunc(); err != nil {
        return item, err
    }
    item = it.items[0]
    it.items = it.items[1:]
    return item, nil
}
```
I would much prefer that.

**Comment by pongad** [comment]
@jba I see what you're saying. PTAL

**Comment by pongad** [comment]
@garrettjonesgoogle Could you take a look since this affects MVVM files?

File: src/main/resources/com/google/api/codegen/go/mock.snip [hunk 2]
```diff
@@ -120,111 +125,181 @@
...
-        return err
+        return nil
     }
 @end
 
-@private testError(test)
-    func {@test.nameWithException}(t *testing.T) {
-        errCode := codes.Internal
-        {@test.mockServiceVarName}.err = grpc.Errorf(errCode, "test error")
+@private testFunc(test)
```
Comment (garrettjonesgoogle) [inline src/main/resources/com/google/api/codegen/go/mock.snip]: this implies a single test function, but there are two - maybe call this `testFunctions`

File: src/main/resources/com/google/api/codegen/go/initcode.snip [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    {@renderResourceName(initValue)}
+  @case "ResourceNameOneofInitValueView"
+    {@initValue.resourceOneofTypeName}.from({@renderResourceName(initValue.specificResourceNameView)})
+  @end
+@end
+
+@private argList(args)
+  @join a : args on ", "
+    {@a}
+  @end
```
Comment (garrettjonesgoogle) [inline src/main/resources/com/google/api/codegen/go/initcode.snip]: The other Go snippets are indented differently - shouldn't this be indented the same as them?

File: src/main/resources/com/google/api/codegen/go/mock.snip (hunk 2 above)
Comment (pongad) [inline src/main/resources/com/google/api/codegen/go/mock.snip]: Done

File: src/main/resources/com/google/api/codegen/go/initcode.snip (hunk 3 above)
Comment (pongad) [inline src/main/resources/com/google/api/codegen/go/initcode.snip]: Changed to 4-space. Thank you for noticing!

**Pull request merged**